workflow/controller/workflowpod.go

```diff
@@ -259,6 +259,14 @@ func (woc *wfOperationCtx) createWorkflowPod(ctx context.Context, nodeName strin
 	// container's PID and root filesystem.
 	pod.Spec.Containers = append(pod.Spec.Containers, mainCtrs...)
 
+	// Add this annotation in order to specify the default container that will be targeted with commands like "kubectl exec/logs"
+	if tmpl.GetType() != wfv1.TemplateTypeContainerSet {
+		pod.ObjectMeta.Annotations[common.AnnotationKeyDefaultContainer] = common.MainContainerName
+	} else {
+		// use last container when pod created from ContainerSet
+		pod.ObjectMeta.Annotations[common.AnnotationKeyDefaultContainer] = pod.Spec.Containers[len(pod.Spec.Containers)-1].Name
```

Comment:
Interesting, not sure what the right logic is here, I think it should be set to the main container if it exists, because that is the only one in a container set that can do some artifacts, otherwise this seems reasonable.